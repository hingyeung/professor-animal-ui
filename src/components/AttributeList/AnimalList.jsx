import React from 'react';
import {
  Link
} from 'react-router-dom';

function AnimalList(props) {
  const animals = props.animals;

  return (
    <div className="container">
      <div className="row">
          <ul className="nav flex-column">
            { animals.map((animal) =>
              <li className="nav-item" key={ animal.name }>
                <Link className={'nav-link'} to={ '/animal/' + animal.id }>{ animal.name }</Link>
              </li>
            ) }
          </ul>
      </div>
    </div>
  )
}

export default AnimalList;