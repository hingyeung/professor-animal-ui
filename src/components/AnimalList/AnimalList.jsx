import React from 'react';
import {
  Link
} from 'react-router-dom';
import AttributeList from 'components/AttributeList';
import PropTypes from 'prop-types';

function AnimalList(props) {
  let compareAnimalName = (a, b) => {
    const lowerCasedNameA = a.name.toLowerCase(),
      lowerCasedNameB = b.name.toLowerCase();

    if (lowerCasedNameA < lowerCasedNameB) {
      return -1;
    } else if (lowerCasedNameA > lowerCasedNameB) {
      return 1;
    } else {
      return 0;
    }
  };

  const animals = props.animals;
  const activeAnimalId = props.activeAnimalId;

  let animalListSortedByName = Object.keys(animals)
    .map((animalId) => animals[animalId])
    .sort(compareAnimalName);

  function isActive(animalId) {
    return animalId === activeAnimalId;
  }

  return (
    <div className="col-12 animal-list-container">
      <ul className="nav nav-pills flex-column">
        <li>
          <Link className="nav-link" to={AttributeList.NEW}>+ Add a new animal</Link>
        </li>
        {
          animalListSortedByName.map(animal => {
            return (
              <li className="nav-item" key={ animal.name }>
                <Link className={ ['nav-link', (isActive(animal.id) ? 'active' : '')].join(' ') }
                      to={ {pathname: '/animal/' + animal.id, state: {activeAnimalId: animal.id}} }>
                  { animal.name }
                </Link>
              </li>
            )
          })
        }
      </ul>
    </div>
  )
}

AnimalList.propTypes = {
  animals: PropTypes.object.isRequired,
  activeAnimalId: PropTypes.string
};

export default AnimalList;