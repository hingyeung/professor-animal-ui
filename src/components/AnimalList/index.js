import AnimalList from './AnimalList.jsx';

export default AnimalList;