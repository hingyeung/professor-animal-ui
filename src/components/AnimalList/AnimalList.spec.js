import React from 'react';
import {shallow} from 'enzyme';
import AnimalList from "./AnimalList";
import {Link} from 'react-router-dom';

describe('AnimalList', function () {
  const animals = {id2: {name: "animal2"}, id1: {name: "animal1"}, id3: {name: "animal3"}};

  it('should render properly with sorted animal names', function () {
    const wrapper = shallow(<AnimalList animals={ animals }/>);
    expect(wrapper.find(Link)).toHaveLength(Object.keys(animals).length + 1);

    const linksToTest = [
      {link: wrapper.find(Link).at(1).prop('children'), name: "animal1"},
      {link: wrapper.find(Link).at(2).prop('children'), name: "animal2"},
      {link: wrapper.find(Link).at(3).prop('children'), name: "animal3"}
    ];

    // https://github.com/airbnb/enzyme/issues/692#issuecomment-262877586
    for (const link of linksToTest) {
      expect(shallow(<div>{ link.link }</div>).text()).toEqual(link.name);
    }

    expect(wrapper).toMatchSnapshot();
  });

  it('should highlight the active animal', function () {
    const wrapper = shallow(<AnimalList animals={ animals } activeAnimalId="id1"/>);
    expect(wrapper).toMatchSnapshot();
  })
});