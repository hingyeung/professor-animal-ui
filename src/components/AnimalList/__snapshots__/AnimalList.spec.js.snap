// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimalList should highlight the active animal 1`] = `
<div
  className="col-12 animal-list-container"
>
  <ul
    className="nav nav-pills flex-column"
  >
    <li>
      <Link
        className="nav-link"
        replace={false}
        to="/new"
      >
        + Add a new animal
      </Link>
    </li>
    <li
      className="nav-item"
      key="animal1"
    >
      <Link
        className="nav-link "
        replace={false}
        to={
          Object {
            "pathname": "/animal/undefined",
            "state": Object {
              "activeAnimalId": undefined,
            },
          }
        }
      >
        animal1
      </Link>
    </li>
    <li
      className="nav-item"
      key="animal2"
    >
      <Link
        className="nav-link "
        replace={false}
        to={
          Object {
            "pathname": "/animal/undefined",
            "state": Object {
              "activeAnimalId": undefined,
            },
          }
        }
      >
        animal2
      </Link>
    </li>
    <li
      className="nav-item"
      key="animal3"
    >
      <Link
        className="nav-link "
        replace={false}
        to={
          Object {
            "pathname": "/animal/undefined",
            "state": Object {
              "activeAnimalId": undefined,
            },
          }
        }
      >
        animal3
      </Link>
    </li>
  </ul>
</div>
`;

exports[`AnimalList should render properly with sorted animal names 1`] = `
<div
  className="col-12 animal-list-container"
>
  <ul
    className="nav nav-pills flex-column"
  >
    <li>
      <Link
        className="nav-link"
        replace={false}
        to="/new"
      >
        + Add a new animal
      </Link>
    </li>
    <li
      className="nav-item"
      key="animal1"
    >
      <Link
        className="nav-link active"
        replace={false}
        to={
          Object {
            "pathname": "/animal/undefined",
            "state": Object {
              "activeAnimalId": undefined,
            },
          }
        }
      >
        animal1
      </Link>
    </li>
    <li
      className="nav-item"
      key="animal2"
    >
      <Link
        className="nav-link active"
        replace={false}
        to={
          Object {
            "pathname": "/animal/undefined",
            "state": Object {
              "activeAnimalId": undefined,
            },
          }
        }
      >
        animal2
      </Link>
    </li>
    <li
      className="nav-item"
      key="animal3"
    >
      <Link
        className="nav-link active"
        replace={false}
        to={
          Object {
            "pathname": "/animal/undefined",
            "state": Object {
              "activeAnimalId": undefined,
            },
          }
        }
      >
        animal3
      </Link>
    </li>
  </ul>
</div>
`;
